merkle-tree/concurrent/tests/tests.rs

```diff
@@ -0,0 +1,729 @@
+            .try_into()
+            .unwrap();
+
+        let changelog_index = merkle_tree.changelog_index();
+        let old_leaf = reference_tree.get_leaf(i);
+        let proof = reference_tree.get_proof_of_leaf(i);
+
+        merkle_tree
+            .update(changelog_index, &old_leaf, &new_leaf, i, proof.as_slice())
+            .unwrap();
+        reference_tree.add_leaf(new_leaf, i);
+    }
+
+    assert_eq!(merkle_tree.current_changelog_index, 5);
+    assert_eq!(merkle_tree.current_root_index, 3);
+}
+
+fn without_changelog<H>()
+where
+    H: Hasher,
+{
+    const HEIGHT: usize = 2;
+    const CHANGELOG: usize = 0;
+    const ROOTS: usize = 8;
+
+    // Our implementation of concurrent Merkle tree.
+    let mut merkle_tree = ConcurrentMerkleTree::<H, HEIGHT, CHANGELOG, ROOTS>::default();
+    merkle_tree.init().unwrap();
+
+    // Reference implementation of Merkle tree which Solana Labs uses for
+    // testing (and therefore, we as well). We use it mostly to get the Merkle
+    // proofs.
+    let leaves = vec![spl_concurrent_merkle_tree::node::EMPTY; 1 << HEIGHT];
+    let mut reference_tree = spl_merkle_tree_reference::MerkleTree::new(leaves.as_slice());
+
+    let mut rng = thread_rng();
+
+    for i in 0..(1 << HEIGHT) {
+        let leaf: [u8; 32] = Fr::rand(&mut rng)
+            .into_bigint()
+            .to_bytes_be()
+            .try_into()
+            .unwrap();
+        merkle_tree.append(&leaf).unwrap();
+        reference_tree.add_leaf(leaf, i);
+    }
+
+    for i in 0..32 {
+        let new_leaf: [u8; 32] = Fr::rand(&mut rng)
+            .into_bigint()
+            .to_bytes_be()
+            .try_into()
+            .unwrap();
+
+        // If `ì` is greater than possible number of leaves, apply a modulus.
+        let i = i % (1 << HEIGHT);
+
+        let changelog_index = merkle_tree.changelog_index();
+        let old_leaf = reference_tree.get_leaf(i);
+        let proof = reference_tree.get_proof_of_leaf(i);
+
+        merkle_tree
+            .update(changelog_index, &old_leaf, &new_leaf, i, proof.as_slice())
+            .unwrap();
+        reference_tree.add_leaf(new_leaf, i);
+    }
+}
+
+#[test]
+fn test_append_keccak() {
+    append::<Keccak>()
+}
+
+#[test]
+fn test_append_poseidon() {
+    append::<Poseidon>()
+}
+
+#[test]
+fn test_append_sha256() {
+    append::<Sha256>()
+}
+
+#[test]
+fn test_update_keccak() {
+    update::<Keccak>()
+}
+
+#[test]
+fn test_update_poseidon() {
+    update::<Poseidon>()
+}
+
+#[test]
+fn test_update_sha256() {
+    update::<Sha256>()
+}
+
+#[test]
+fn test_overfill_tree_keccak() {
+    overfill_tree::<Keccak>()
+}
+
+#[test]
+fn test_overfill_tree_poseidon() {
+    overfill_tree::<Poseidon>()
+}
+
+#[test]
+fn test_overfill_tree_sha256() {
+    overfill_tree::<Sha256>()
+}
+
+#[test]
+fn test_overfill_changelog_keccak() {
+    overfill_changelog_and_roots::<Keccak>()
+}
+
+#[test]
+fn test_without_changelog_keccak() {
+    without_changelog::<Keccak>()
```

Comment:
I like your test naming, it makes it very easy to see which properties you are testing!